docs/flags.md

```diff
@@ -0,0 +1,130 @@
+
+flags.exclude_all_resources()
+for r in cfg['to-run']:
+  flags.include_resource(r)
+```
+
+### Run a defined set of services
+Your app has many services, call them A, B, C, and D. You have defined modes you want users to be able to start (e.g. consumer vs. enterprise) and you don't want them to have to remember which service is necessary for which mode.
+
+You want to be able to write a doc that tells users to run `tilt up consumer` the first time and know that's enough that they'll be set up. Running Tilt subsequent times without arguments should default to the mode they already have set.
+
+#### Command-Lines
+* `tilt up`: run all services
+* `tilt up consumer`: run consumer services and record mode
+* `tilt up` (a second time): run consumer services
+
+#### Tiltfile
+```python
+flags.define_string_list("to-run", args=True)
+cfg = flags.parse()
+
+groups = {
+  'consumer': ['a', 'b', 'c'],
+  'enterprise': ['a', 'b', 'd'],
+}
+
+
+
+
+
+
+
+
+
+
+
+
+flags.exclude_all_resources()
+for arg in cfg['args']:
+  if arg in groups:
+    for r in groups[arg]:
+      flags.include_resource(r)
+  else:
+    flags.include_resource(r)
+```
+
+By making the flag sticky (the default), Tilt will record the flag between runs.
+
+
+### Specify services to edit
+You have many services that your users frequently need to run, but don't expect to edit. Your YAML includes images from your CI that are suitably recently (and match what's in production or staging).
+
+#### Command-Lines
+* `tilt up`: run all services, editing none
+* `tilt up -- --to-edit b`: run all services, editing b
+* `tilt up -- consumer --to-edit b --to-edit c`: run consumer services and setup b and c for editing.
+
+#### Tiltfile
+```python
+flags.define_string_list("to-run", args=True)
+flags.define_string_list("to-edit")
+cfg = flags.parse()
+
+# omitting the code from above Tiltfile that includes correct resources
+
+# only configure the build for services we want to edit
+if 'a' in cfg['to-edit']:
+  docker_build('a', './a')
+if 'b' in cfg['to-edit']:
+  docker_build('b', './b')
+if 'c' in cfg['to-edit']:
+  docker_build('c', './c')
+if 'd' in cfg['to-edit']:
+  docker_build('d', './d')
+```
+
+## How Flags Work
+Tilt stores flags in a file next to the Tiltfile called `tilt_flags.json`. When the Tiltfile executes `flags.parse()`, it loads that file as JSON. If it's the first execution after a Tilt up, the execution will merge the flags passed on the command-line with the values from the file (using the definition of flags as a guide) and write the new values to the file if necessary.
+
+Any change to the flags file will cause Tilt to reexecute the Tiltfile. You can modify the flags file with an editor or write scripts that write values into it, and Tilt will update.
+
+A call to `tilt flags` will connect to the running Tilt instance on the server port and cause Tilt to reexecute the Tiltfile, using the passed command-line args as if it's the first execution after a Tilt up.
```

Comment:
At the risk of getting nitty, for both this and the line above - I might say "any change to flags will cause Tilt to reexecute the Tiltfile", and then the other mechanisms might change flags.

Which is to say, maybe some operation touches the flags file but doesn't change its contents, or I run `tilt flags` to see the current values but don't change anything, and that shouldn't necessarily run the Tiltfile.

Reply:
All of our other watching is based on timing, not on contents.

I agree that tilt flags --help shouldn't reinvoke.

Reply:
> Think I mentioned this above, but this seems like a risky assumption (i.e. that there's only one running Tilt instance)

Risky, but probably fine for now? Like, presumably a vanishingly small fraction of our users have multiple tilts, and also a relatively small fraction of our users will be using this, and we can solve that when people complain?

at some point we might want `screen(1)`-like behavior for tilt instances
i.e., screen is able to tell how many screens are running - if there's exactly one, it uses that. if there are multiple, it lists them all and tells you how to specify which one you want

Reply:
> All of our other watching is based on timing, not on contents.

I don't think this decision matters much either way, but fwiw I was mostly thinking of the analogue to manifest configs - if I touch a yaml file, the Tiltfile will reexecute, re-load the yaml contents into the manifest definition, observe that the manifest definition has not changed, and will not trigger a resource build, which to a user, kind of ends up feeling like watching based on contents, depending on how disruptive that project's Tiltfile execution is.